notes/notes.go

```diff
@@ -0,0 +1,345 @@
+}
+
+func (p *parsedNotes) edit(index int, note string, author *discordgo.User) error {
+	if len(p.notes) <= index {
+		// this shouldn't be possible
+		logger.WithField("guild", p.guildID).WithField("index", index).WithField("length", len(p.notes)).Warn("a custom ID tried to select a nonexistent note")
+		return invalidSelectionErr
+	}
+	note = strings.TrimSpace(note)
+	lines := strings.Split(note, "\n")
+	now := time.Now()
+	p.notes[index] = &Note{
+		noteLines: lines,
+		signature: author.String(),
+		authorID:  author.ID,
+		updatedAt: now,
+	}
+	return nil
+}
+
+func (p *parsedNotes) delete(index int) error {
+	if len(p.notes) <= index {
+		// this shouldn't be possible
+		logger.WithField("guild", p.guildID).WithField("index", index).WithField("length", len(p.notes)).Warn("a custom ID tried to select a nonexistent note")
+		return invalidSelectionErr
+	}
+	p.notes[index] = &Note{}
+	return nil
+}
+
+func (p *parsedNotes) save(ctx context.Context) error {
+	var formattedNotes []string
+	for _, n := range p.notes {
+		formattedNotes = append(formattedNotes, n.formatNote())
+	}
+
+	logger.Info(formattedNotes)
+
+	model := models.Note{
+		GuildID:  p.guildID,
+		MemberID: p.userID,
+		Notes:    formattedNotes,
+	}
+
+	whitelist := boil.Whitelist("notes")
+	return model.Upsert(ctx, common.PQ, true, []string{"guild_id", "member_id"}, whitelist, boil.Infer())
+}
+
+const noNotesMsg = "No shared staff notes written for this user."
+
+func (p *parsedNotes) createMessageContent() *discordgo.MessageSend {
+	msg := &discordgo.MessageSend{}
+	if len(p.notes) == 0 {
+		msg.Content = noNotesMsg
+		return msg
+	}
+
+	for i := range maxNotesPerUser {
+		if !(len(p.notes) > i && len(p.notes[i].noteLines) > 0 && p.notes[i].noteLines[0] != "") {
+			continue
+		}
+
+		selectedNote := p.notes[i]
+		msg.Embeds = append(msg.Embeds, &discordgo.MessageEmbed{
+			Title:       fmt.Sprint("Note #", i+1),
+			Description: strings.Join(selectedNote.noteLines, "\n"),
+			Timestamp:   selectedNote.updatedAt.Format(time.RFC3339),
+			Footer:      &discordgo.MessageEmbedFooter{Text: "by " + selectedNote.signature},
+		})
+	}
+
+	if len(msg.Embeds) < 1 {
+		msg.Content = noNotesMsg
+		return msg
+	}
+
+	name := strconv.FormatInt(p.userID, 10)
+	ms, _ := bot.GetMember(p.guildID, p.userID)
+	if ms != nil {
+		name = ms.User.String()
+	}
+	msg.Content = "## Shared Staff Notes for " + name
+	return msg
+}
+
+type notesButtonType int
+
+const (
+	notesButtonTypeNew notesButtonType = iota
+	notesButtonTypeEdit
+	notesButtonTypeDelete
+)
+
+func (p *parsedNotes) customID(index int, buttonType notesButtonType) string {
+	args := []string{strconv.FormatInt(p.userID, 10)}
+	args = append(args, strconv.Itoa(index))
+	switch buttonType {
+	case notesButtonTypeNew:
+		args[1] = "new"
+	case notesButtonTypeEdit:
+		args = append(args, "edit")
+	case notesButtonTypeDelete:
+		args = append(args, "delete")
+	}
+
+	id := strings.Join(args, "-")
+	id = "notes_" + id
+	return id
+}
```

Comment:
This code for generating the custom ID is tightly coupled to the code that decodes it (handleNoteButton and friends), but they are rather far apart in the source. What do they think about adding a function that parses data out of the custom ID right beside this function, and using it in handleNoteButton? Something like

```go
type NoteAction {
    type notesButtonType // a better name for the type might be noteActionType 
    index int // unset if action==new
}

func encodeCustomID(action noteAction, userID int64) string // replaces parsedNotes.customID
func decodeCustomID(customID string) (NoteAction, bool)    // use in handleNoteAction and friends
```

Reply:
Oo, I like it. Will work on this later.